File: DSA/DynamicProgramming/Code/LCS.py
# LCS (DP) in Python

def lcs(X, Y):
    m = len(X)
    n = len(Y)
    dp = [[0]*(n+1) for _ in range(m+1)]

    # Build DP table
    for i in range(m+1):
        for j in range(n+1):
            if i == 0 or j == 0:
                dp[i][j] = 0
            elif X[i-1] == Y[j-1]:
                dp[i][j] = 1 + dp[i-1][j-1]
            else:
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])

    # Reconstruct LCS string
    i, j = m, n
    lcs_str = []
    while i > 0 and j > 0:
        if X[i-1] == Y[j-1]:
            lcs_str.append(X[i-1])
            i -= 1
            j -= 1
        elif dp[i-1][j] > dp[i][j-1]:
            i -= 1
        else:
            j -= 1

    lcs_str.reverse()  # reverse to get correct order
    return dp[m][n], "".join(lcs_str)


X = "AGGTAB"
Y = "GXTXAYB"

print("=== LCS (Python) ===\n")
length, subsequence = lcs(X, Y)
print(f"Length of LCS: {length}")
print(f"LCS: {subsequence}")
print("\nTime Complexity: O(mn)")
print("Space Complexity: O(mn)")

File: DSA/DynamicProgramming/Code/LCS.java
// LCS (DP) in Java

public class LCS {
    static String lcs(String X, String Y) {
        int m = X.length();
        int n = Y.length();
        int[][] dp = new int[m+1][n+1];

        // Build DP table
        for (int i = 0; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                if (i == 0 || j == 0)
                    dp[i][j] = 0;
                else if (X.charAt(i-1) == Y.charAt(j-1))
                    dp[i][j] = 1 + dp[i-1][j-1];
                else
                    dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
            }
        }

        // Reconstruct LCS string
        int i = m, j = n;
        StringBuilder lcsStr = new StringBuilder();
        while (i > 0 && j > 0) {
            if (X.charAt(i-1) == Y.charAt(j-1)) {
                lcsStr.append(X.charAt(i-1));
                i--;
                j--;
            } else if (dp[i-1][j] > dp[i][j-1]) {
                i--;
            } else {
                j--;
            }
        }

        return lcsStr.reverse().toString(); // reverse to get correct order
    }

    public static void main(String[] args) {
        System.out.println("=== LCS (Java) ===\n");
        String X = "AGGTAB";
        String Y = "GXTXAYB";

        String subsequence = lcs(X, Y);
        System.out.println("Length of LCS: " + subsequence.length());
        System.out.println("LCS: " + subsequence);

        System.out.println("\nTime Complexity: O(mn)");
        System.out.println("Space Complexity: O(mn)");
    }
}